compiler/circle-tensordump/src/Dump.cpp

```diff
@@ -0,0 +1,345 @@
+        os << std::endl;
+        os << " " + print_format2 + "   └── zero_point : ";
+        ::print_comma_sepearted(os, zero_point);
+        os << std::endl;
+      }
+
+      // buffer
+      print_buffer(os, buff_idx, buff_data_ptr, tensor->type());
+      os << std::endl;
+    }
+  }
+}
+
+} // namespace circletensordump
+
+namespace
+{
+
+// HDF5 forbids the inclusion of '/' in the name
+std::string mangle(const std::string &name)
+{
+  std::string ret{name};
+  std::replace(ret.begin(), ret.end(), '/', '_');
+  return ret;
+}
+
+H5::PredType hdf5_dtype_cast(const circle::TensorType &circle_type)
+{
+  switch (circle_type)
+  {
+    case circle::TensorType_UINT8:
+    {
+      return H5::PredType::NATIVE_INT8;
+    }
+    case circle::TensorType_INT32:
+    {
+      return H5::PredType::NATIVE_INT32;
+    }
+    case circle::TensorType_INT64:
+    {
+      return H5::PredType::NATIVE_INT64;
+    }
+    case circle::TensorType_FLOAT32:
+    {
+      ;
+      return H5::PredType::NATIVE_FLOAT;
+    }
+    default:
+      throw std::runtime_error("NYI tensor type : " + std::to_string(circle_type));
+  }
+}
+
+template <typename T>
+void write_data_to_hdf5(H5::H5File &file, const flatbuffers::Vector<T> *vec,
+                        const H5::PredType &type, std::string &group_name, std::string dataset_name)
+{
+  if (vec == nullptr)
+    return;
+  const uint32_t rank = vec->size();
+
+  // set dataset dimension
+  std::vector<hsize_t> dims(rank);
+  for (uint32_t d = 0; d < rank; d++)
+  {
+    dims[d] = vec->Get(d);
+  }
+  auto dataspace = std::make_unique<H5::DataSpace>(rank, dims.data());
+  auto dataset = std::make_unique<H5::DataSet>(
+      file.createDataSet(group_name + "/" + dataset_name, type, *dataspace));
+  dataset->write(vec->data(), type);
+}
+
+} // namespace
+
+namespace circletensordump
+{
+
+/**
+ *  HDF5 layout is like below
+ *
+ *  Group "/"
+ *   ㄴGroup "tensor name"
+ *     ㄴDataset "weights"    : Shape (x, y, ...), type(uint8, int16)
+ *     ㄴDataset "min"        : Shape (n)
+ *     ㄴDataset "max"        : Shape (n)
+ *     ㄴDataset "scale"      : Shape (m)
+ *     ㄴDataset "zero_point" : Shape (m)
+ *
+ *  NOTE All Dataset is optional. It means that if tensor doesn't have the data, it won't be created
+ * as a Dataset
+ *
+ */
+
+void DumpTensorsToHdf5::run(std::ostream &os, const circle::Model *model,
+                            const std::string &output_path)
+{
+  // loads a circle model
+  circletensordump::Reader reader(model);
+  uint32_t num_subgraph = reader.num_subgraph();
+
+  // create a hdf5 file
+  H5::H5File file{output_path, H5F_ACC_TRUNC};
+
+  for (uint32_t subgraph_idx = 0; subgraph_idx < num_subgraph; subgraph_idx++)
+  {
+    reader.select_subgraph(subgraph_idx);
+  }
+
+  auto tensors = reader.tensors();
+  for (const auto &tensor : *tensors)
+  {
+    // create a group for each tensor whose name is its tensor name
+    std::string group_name = ::mangle(::mangle(tensor->name()->c_str()));
```

Comment:
Why do you mangle twice?

Reply:
Maybe I got possessed by "off" whose surname is "work"ㅠ_ㅠ